### PR TITLE
Alternative version of #2999: Add more tests on using sbt as the build server

### PR DESCRIPTION
**Pull request by adpi2**
Alternative version of #2999 where `BuildServerInitializer` and `BuildToolLayout` are passed as parameters rather than being mixed as traits.

@tgodzik @ckipp01 do you think this one is better than the previous one?

### CODE REVIEW

File: tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala [hunk 1]
```diff
@@ -33,23 +32,24 @@ class SbtServerSuite
...
       _ = assertNoDiff(
         client.workspaceMessageRequests,
         List(
-          importBuildMessage,
-          progressMessage
+          importBuildMessage
         ).mkString("\n")
       )
       _ = client.messageRequests.clear()
-      _ = assertStatus(_.isInstalled)
```
Comment (tgodzik) [inline tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala:52]: Why remove this check?

File: tests/unit/src/main/scala/tests/BuildServerInitializer.scala [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      _ = client.selectBspServer = { _ => new MessageActionItem("sbt") }
+      _ <- server.executeCommand(ServerCommands.BspSwitch.id)
+    } yield {
+      if (!expectError) {
+        server.assertBuildServerConnection()
+      }
+    }
+  }
+
+  private def generateBspConig(workspace: AbsolutePath): Unit = {
```
Comment (tgodzik) [inline tests/unit/src/main/scala/tests/BuildServerInitializer.scala:106]:
```suggestion
  private def generateBspConfig(workspace: AbsolutePath): Unit = {
```

File: tests/unit/src/main/scala/tests/BuildServerInitializer.scala [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    val bspFolder = workspace.resolve(".bsp")
+    val sbtJson = bspFolder.resolve("sbt.json")
+    // don't overwrite existing BSP config
+    if (!sbtJson.isFile) {
+      // we create bsp/sbt.json file manually because `sbt bspConfig` takes too long
+      val sbtLaunchJar =
```
Comment (tgodzik) [inline tests/unit/src/main/scala/tests/BuildServerInitializer.scala:112]: Are you able to just use `SbtBuildTool` and run it?

File: tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala [hunk 4]
```diff
@@ -33,23 +32,24 @@ class SbtServerSuite
...
-            |scalaVersion := "${V.scala212}"
-            |""".stripMargin
-      )
+      _ <- server.initialize()
+      _ <- server.initialized()
       _ = assertNoDiff(
         client.workspaceMessageRequests,
         List(
-          importBuildMessage,
-          progressMessage
```
Comment (tgodzik) [inline tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala:48]: Why did the progressMessage disappear ? It's better, I am just asking out of curiosity.

**Review by tgodzik — APPROVED** [review]
LGTM! I left just some minor comments. Thanks a lot for the follow up!

File: tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala (hunk 4 above)
Comment (adpi2) [reply tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala:48]: In `TestingClient` the default response of the `importBuildMessage` is:
```scala
var importBuild: MessageActionItem = ImportBuild.notNow
```

So Metals does not run `bloopInstall` and that's why there is no `progressMessage` anymore. Instead it switches to sbt as the build server.

File: tests/unit/src/main/scala/tests/BuildServerInitializer.scala (hunk 3 above)
Comment (adpi2) [reply tests/unit/src/main/scala/tests/BuildServerInitializer.scala:112]: That's tested in `SbtServerSuite.generate`. I am not using `SbtBuildTool` here to save some time by not running `sbt bspConfig`.

File: tests/unit/src/main/scala/tests/BuildServerInitializer.scala (hunk 2 above)
Comment (adpi2) [reply tests/unit/src/main/scala/tests/BuildServerInitializer.scala:106]: :+1:

File: tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala (hunk 1 above)
Comment (adpi2) [reply tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala:52]: All tests used to run `bloopInstall` but that takes too long and it is not always useful. Here we want to test sbt server as the build server so we don't need to run `bloopInstall`, it is already tested in `SbtBloopLspSuite`.

**Pull request merged**